dali/operators/math/expressions/broadcasting.cc

```diff
@@ -274,33 +274,165 @@ void ExpandToNDims(TensorShape<> &sh, int ndim) {
+  auto all_ones = [=](int dim) {
+    for (int k = 0; k < shapes.size(); k++) {
+      if (1 != (*shapes[k])[dim]) {
+        return false;
+      }
     }
-    int j = i;
-    for (; j < full_ndim; j++) {
-      if (lhs[j] != rhs[j]) break;
+    return true;
+  };
+
+  // True if all dim-th extents are the same, or equal to one
+  auto all_same_or_one = [=](int dim) {
+    int64_t extent = -1;
+    for (int k = 0; k < shapes.size(); k++) {
+      int64_t value = (*shapes[k])[dim];
+      if (value == 1)
+        continue;
+      if (extent == -1) {
+        extent = value;
+      } else if (extent != value) {
+        return false;
+      }
     }
-    if (i < j) {
-      group_dims.emplace_back(i, j - i);
+    return true;
+  };
+
+  // Can collapse dimensions with ones
+  // when there is no transition from or to an 'odd' one
+  // (meaning that the rest of the extents are not all one
+  // in the two dimensions)
+  // Examples:
+  // 1. Can collapse even if we have transition from 2 to 1,
+  //    because there are only ones on the second dimension
+  // {2 1} -> {2}
+  // {1 1} -> {1}
+  // 2. Can NOT collapse because there is a transition to 1
+  // {2 1} -> {2 1}
+  // {2 2} -> {2 2}
+  auto can_merge_ones = [=](int dim) {
+    for (int k = 0; k < shapes.size(); k++) {
+      auto &sh = *shapes[k];
+      if ((sh[dim] != 1 && sh[dim + 1] == 1 && !all_ones(dim + 1)) ||
+          (sh[dim] == 1 && !all_ones(dim) && sh[dim + 1] != 1)) {
+        return false;
+      }
     }
-    i = j;
+    return true;
+  };
+
+  // We can merge a dimension with the next if any of the following conditions apply:
+  // 1. All extents on that dimension and the next are the same (that also includes all ones)
+  // 2. All extents on that dimension and the next are the same or one,
+  //    and the ones are either present or not present in both dimensions
+  auto can_merge_next = [=](int dim) {
+    if (all_same(dim) && all_same(dim + 1))
+      return true;
+    else if (all_same_or_one(dim) && all_same_or_one(dim + 1) && can_merge_ones(dim))
+      return true;
+    else
+      return false;
+  };
+
+  // Updates the volumes with the new dimension and returns true if the resulting volumes are
+  // compatible (all the same except for 1s)
+  auto compatible_vol = [=](SmallVector<int64_t, kMaxArity>& volumes, int dim) {
+    int64_t curr_vol = -1;
+    for (int k = 0; k < shapes.size(); k++) {
+      volumes[k] *= (*shapes[k])[dim];
+      if (volumes[k] == 1)
+        continue;  // always compatible
+      else if (curr_vol == -1)
+        curr_vol = volumes[k];
+      else if (curr_vol != volumes[k])
+        return false;  // not compatible
+    }
+    return true;
+  };
+
+  int i = 0;
+  SmallVector<int64_t, kMaxArity> volumes;
+  volumes.resize(shapes.size(), 1);
+  for (; i < full_ndim - 1;) {
+    if (compatible_vol(volumes, i)) {
+      int j = i + 1;
+      for (; j < full_ndim;) {
+        if (compatible_vol(volumes, j) && can_merge_next(j - 1)) {
```

Comment:
Why do we need both `compatible_vol` and `can_merge_next`? Isn't `can_merge_next` enough?